pt2pt: add MPID_Allocate_vci

hzhou opened the pull request:
## Pull Request Description
Reserved vci can be used to isolate communications. For example, stream-based progress should not involve GPU path. Using a reserved vci that normal GPU traffic won't touch can ensure that. Similarly, we can use a reserved vci for dynamic process connections.

Since users always know exactly their thread mapping, it may be simpler and more reliable to let user directly specify vci, rather than we simplicity do hashing. To allow such explicit vci extensions, we need a way to pass down the vci information to device layer. Extending the `attr` parameter can achieve that.

[skip warnings]
## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

hzhou commented:
test:mpich/custom --with-ch4-reserved-vcis=2
netmod: ch4:ofi

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
With MPIX Stream Communicator, since the information is in the communicator, we don't need change the ADI nor netmod API to pass the explicit vci information down. Thus will remove the last two commits.

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

raffenet reviewed (commented):
I'm a little confused by this statement

> VCI 0 is the default global vci, that is always used for implicit vci hashing.

I believe we're moving to passing explicit VCIs thru the ADI in some cases, but I would think that in the implicit case the VCI information is ignored by the device.

hzhou commented:
> I'm a little confused by this statement
> 
> > VCI 0 is the default global vci, that is always used for implicit vci hashing.
> 
> I believe we're moving to passing explicit VCIs thru the ADI in some cases, but I would think that in the implicit case the VCI information is ignored by the device.

I updated the commit to have `MPID_Allocate_vci` return explicit error code (and use vci pointer for value out). Using `0` to indicate error is a bit hacky. The corresponding commit message is now removed.

hzhou commented on src/mpid/ch4/src/ch4_init.c:
```diff
@@ -414,16 +414,17 @@ int MPID_Init(int requested, int *provided)
-        MPIR_Assert(MPIDI_global.n_vcis <= MPIR_REQUEST_NUM_POOLS);
-    }
+    MPIR_Assert(MPIR_CVAR_CH4_NUM_VCIS >= 1);   /* maximum number vcis can be reserved */
+    MPIR_Assert(MPIR_CVAR_CH4_RESERVE_VCIS >= 0);       /* maximum number vcis can be reserved */
```
I updated the comment for `MPIR_CVAR_CH4_NUM_VCIS`. That was the original comment, but now we changed its meaning to number of implicit vcis rather than maximum number of vcis.

raffenet commented on src/mpi/errhan/errnames.txt:
```diff
@@ -949,6 +949,11 @@ is too big (> MPIU_SHMW_GHND_SZ)
 **set_thread_affinity %d:Failed to set the async thread affinity to the logical processor [%d]
 
+## MPIX_Stream
+**ch3nostream:Stream is not supported in ch3.
+**ch4nostream:No streams available. Configure --enable-thread-cs=per-vci and --with-ch4-max-vcis=# to enable streams.
+**outofstream:No streams available. Use MPIR_CVAR_CH4_RESERVE_VCIS to reserve the number of streams can be allocated.
```
nitpick: its not obvious that vci is what is meant by "stream" in these error messages

hzhou replied:
What do you suggest?

raffenet replied:
I guess the issue is that this error message is meant for callers of `MPIX_Stream_create`? In that case it makes sense, its just odd to try to allocate X and get an error message that there are no more Y.

hzhou replied:
I see. Right, the message is meant for `MPIX_Stream_create`. Only the lower layer has the details on what is wrong, so we have to craft the message there.
